### PR TITLE
[Feature] Add MosaicMixupDataset

### PR DESCRIPTION
**Pull request by HAOCHENYE**
- Update MosaicMixupDataset
- Formatting codes
- Add doc string

### CODE REVIEW

File: mmdet/datasets/dataset_wrappers.py [hunk 1]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
     def __len__(self):
         """Length after repetition."""
         return len(self.repeat_indices)
+
+@DATASETS.register_module()
```
Comment (ZwwWayne) [inline mmdet/datasets/dataset_wrappers.py:288]: Need two lines above

File: mmdet/datasets/dataset_wrappers.py [hunk 2]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+    - The target of mixup transform is the weighted average of mixup image and mosaic image.
+
+
+    dataset: Custom dataset with LoadImageFromFile and LoadAnnotations pipeline
+    mosaic_pipeline: default RandomAffine, applied after mosaic before mixup
+    pipeline: custom pipeline after mosaic and mixup transform
+
+   Args:
+       dataset (CunstomDataset): Child of CunstomDataset which can get image and annotations by index.
+       mosaic_pipeline (pipeline): Augmentations after mosaic
```
Comment (ZwwWayne) [inline mmdet/datasets/dataset_wrappers.py:357]: pipeline -> dict

File: mmdet/datasets/dataset_wrappers.py [hunk 3]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+    - Another random image is picked by dataset  and embedded in the top left patch(after padding and resizing)
+    - The target of mixup transform is the weighted average of mixup image and mosaic image.
+
+
+    dataset: Custom dataset with LoadImageFromFile and LoadAnnotations pipeline
+    mosaic_pipeline: default RandomAffine, applied after mosaic before mixup
+    pipeline: custom pipeline after mosaic and mixup transform
+
+   Args:
+       dataset (CunstomDataset): Child of CunstomDataset which can get image and annotations by index.
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:356]: :obj:`CunstomDataset`

File: mmdet/datasets/dataset_wrappers.py [hunk 4]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+       mosaic_pipeline (pipeline): Augmentations after mosaic
+       pipeline (pipeline): Augmentations after mosaic and mixup.
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:358]: Please add default value for these arguments

File: mmdet/datasets/dataset_wrappers.py [hunk 5]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
     def __len__(self):
         """Length after repetition."""
         return len(self.repeat_indices)
+
+@DATASETS.register_module()
+class MosaicMixUpDataset:
+    """Mosaic and MixUp transform used in YOLOX
+
+       MosaicMixUpDataset first combime four images from CustomDataset,
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:292]:
```
"""Mosaic and MixUp transform used in YOLOX

MosaicMixUpDataset first combime four images from CustomDataset,
```
No indentation for detailed doc

Comment (jshilong) [reply mmdet/datasets/dataset_wrappers.py:292]: combimes

File: mmdet/datasets/dataset_wrappers.py [hunk 6]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+       and then mixup the output with another random image. This mosaic
+       logic is consistent with yolov5 and YOLOX. MixUp output will be
+       further random resized each 10(default) iters. The dynamic_scale,
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:295]: every 10(default) iters

File: mmdet/datasets/dataset_wrappers.py [hunk 7]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+       enable_mosaic, mixup_scale attributions are controlled by
+       YoloXProcessHook.
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:297]: Can we add more detailed doc for each attribute is modified by which hook

File: mmdet/datasets/dataset_wrappers.py [hunk 8]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+               |      |                 |     |
+               |      |-----------------+     |
+               |             pad              |
+               +------------------------------+
+    The mixup transform step:
+    - Another random image is picked by dataset  and embedded in the top left patch(after padding and resizing)
+    - The target of mixup transform is the weighted average of mixup image and mosaic image.
+
+
+    dataset: Custom dataset with LoadImageFromFile and LoadAnnotations pipeline
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:351]: Seems we need remove these comments

File: mmdet/datasets/dataset_wrappers.py [hunk 9]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+                 mixup_flip_ratio=0.5,
+                 pad_value=114):
+        self.dataset = dataset
+        self.CLASSES = dataset.CLASSES
+        if hasattr(self.dataset, 'flag'):
+            self.flag = dataset.flag
+        self.mosaic_pipeline = Compose(mosaic_pipeline) if mosaic_pipeline is not None else None
+        self.mixup_pipeline = Compose(mixup_pipeline) if mixup_pipeline is not None else None
+        self.pipeline = Compose(pipeline)
+        self.enable_mosaic = enable_mosaic
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:386]: It is recommended to add an interface for modifying the value, directly change the value of an attribute from outside is non-compliant.
Same comments for other attributes which would be modified by hook

File: mmdet/datasets/dataset_wrappers.py [hunk 10]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+
+        results['img'] = mosaic_img.astype(np.float32)
+        results['img_shape'] = mosaic_img.shape
+        results['ori_shape'] = mosaic_img.shape
+        results['gt_bboxes'] = mosaic_bboxes
+        results['gt_labels'] = mosaic_labels
+
+        return results
+
+    def _mosiac_combine(self, loc, center_position_xy, img_shape_wh):
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:480]: docstr here

File: mmdet/datasets/dataset_wrappers.py [hunk 11]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+            mixup_gt_labels = np.concatenate(
+                (results['gt_labels'], retrieve_gt_labels), axis=0)
+
+            results['img'] = mixup_img
+            results['gt_bboxes'] = mixup_gt_bboxes
+            results['gt_labels'] = mixup_gt_labels
+
+        return results
+
+    def filter_box_candidates(self, box1, box2, wh_thr=2, ar_thr=20, area_thr=0.2):
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:616]: ar_thr may be confusing

File: mmdet/datasets/dataset_wrappers.py [hunk 12]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+        """ Compute candidate boxes which include follwing 5 things:
+        box1 before augment, box2 after augment, wh_thr (pixels), aspect_ratio_thr, area_ratio
+        """
+        w1, h1 = box1[2] - box1[0], box1[3] - box1[1]
+        w2, h2 = box2[2] - box2[0], box2[3] - box2[1]
+        ar = np.maximum(w2 / (h2 + 1e-16), h2 / (w2 + 1e-16))  # aspect ratio
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:622]: make `1e-16` as an argument instead of hardcode

File: mmdet/datasets/dataset_wrappers.py (hunk 11 above)
Comment (jshilong) [reply mmdet/datasets/dataset_wrappers.py:616]: Please add docstr for this function

File: mmdet/datasets/dataset_wrappers.py [hunk 13]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+        # 3. scale jit
+        scale_ratio *= jit_factor
+        out_img = mmcv.imresize(out_img, (int(out_img.shape[1] * jit_factor), int(out_img.shape[0] * jit_factor)))
+
+        # 4. flip
+        if is_filp:
+            out_img = out_img[:, ::-1, :]
+
+        # 5. random crop
+        origin_img = results['img']
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:570]: Seems usually use `ori` as the prefix is enough

File: mmdet/datasets/dataset_wrappers.py [hunk 14]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+               |    |  cropped    |           |
+               |pad |   image3    |  image4   |
+               |    |             |           |
+               +----|-------------+-----------+
+                    |             |
+                    +-------------+
+
+    The mosaic transform step:
+
+    - choose the mosaic center as the intersection  of 4 images
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:320]: `intersection of 4 images`
remove the extra space

File: mmdet/datasets/dataset_wrappers.py [hunk 15]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+    - Get the left top image according to the index given by dataloader,
+      and ramdomly sample another 3 images from the custom dataset.
+    - sub image will be resized every 10 iters controlled by YoloXProcessHook
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:323]: Please change `YoloXProcessHook` to the real name of corresponding hook

File: mmdet/datasets/dataset_wrappers.py [hunk 16]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+    - sub image will be cropped if image is larger than mosaic patch
+    - mosaic and mixup transform will be disabled last several epochs
+      controlled by YoloXProcessHook
+    - mosaic image will be futher augment by mosaic pipeline(default RandomAffine)
+    - mosaic and mixup image will be futher augment by pipeline
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:328]: mosaic and mixup images will be further augmented by pipeline

File: mmdet/datasets/dataset_wrappers.py [hunk 17]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+    The mosaic transform step:
+
+    - choose the mosaic center as the intersection  of 4 images
+    - Get the left top image according to the index given by dataloader,
+      and ramdomly sample another 3 images from the custom dataset.
+    - sub image will be resized every 10 iters controlled by YoloXProcessHook
+    - sub image will be cropped if image is larger than mosaic patch
+    - mosaic and mixup transform will be disabled last several epochs
+      controlled by YoloXProcessHook
+    - mosaic image will be futher augment by mosaic pipeline(default RandomAffine)
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:327]: mosaic and mixup images will be further augmented by pipeline

File: mmdet/datasets/dataset_wrappers.py [hunk 18]
```diff
@@ -280,3 +284,343 @@ def __getitem__(self, idx):
...
+                results = copy.deepcopy(self.dataset[indices[i - 1]])
+
+            img_i = results['img']
+            h_i, w_i = img_i.shape[:2]
+            # keep_ratio resize
+            scale_ratio_i = min(self.img_scale[0] / h_i, self.img_scale[1] / w_i)
+            img_i = mmcv.imresize(img_i, (int(w_i * scale_ratio_i), int(h_i * scale_ratio_i)))
+
+            # compute the combine parameters
+            paste_coord, crop_coord = self._mosiac_combine(loc, center_position, img_i.shape[:2][::-1])
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:446]: Seem we should find a more compatible name for `_mosiac_combine`

File: mmdet/datasets/pipelines/transforms.py [hunk 19]
```diff
@@ -1903,3 +1902,393 @@ def __repr__(self):
...
+        for i in range(15):
+            index = random.randint(0, len(dataset))
+            gt_bboxes_i = dataset.get_ann_info(index)['bboxes']
+            if len(gt_bboxes_i) != 0:
+                break
+
+        if len(gt_bboxes_i) == 0:
+            return []
+        else:
+            return random.randint(0, len(dataset))
```
Comment (RangiLyu) [inline mmdet/datasets/pipelines/transforms.py:2161]: Why not return a list?

File: mmdet/datasets/dataset_wrappers.py [hunk 20]
```diff
@@ -1,282 +1,336 @@
...
+            self.flag = dataset.flag
+        self.num_sample = len(dataset)
+
+        self.dynamic_scale = dynamic_scale
+        self.enable_mixaug = enable_mixaug
+
+    def disable_mixaug(self):
+        for pipeline in self.pipelines:
+            if hasattr(pipeline, "get_indexes"):
+                self.pipelines.remove(pipeline)
```
Comment (RangiLyu) [inline mmdet/datasets/dataset_wrappers.py:316]: If remove the pipeline, It cannot be turn on later. What about skipping the pipeline which has attribute "get_indexes" in __getitem__()?

Comment (hhaAndroid) [reply mmdet/datasets/dataset_wrappers.py:316]: Can’t write like that, it’s not flexible enough

File: mmdet/datasets/pipelines/transforms.py [hunk 21]
```diff
@@ -1903,3 +1921,454 @@ def __repr__(self):
...
+        repr_str += f'center_scale_ratio={self.center_scale_ratio})'
+        repr_str += f'pad_value={self.pad_value})'
+        return repr_str
+
+
+@PIPELINES.register_module()
+class MixUp:
+    """MixUp data augmentation.
+
+                         mixup transform
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:2167]: Use code-block to make it normally show in sphinx.

File: mmdet/datasets/pipelines/transforms.py [hunk 22]
```diff
@@ -1903,3 +1921,454 @@ def __repr__(self):
...
+        """Call function to collect indexes.
+
+        Args:
+            dataset (:obj:`CustomDataset`): The dataset.
+
+        Returns:
+            list: indexes.
+        """
+
+        for i in range(15):
```
Comment (ZwwWayne) [inline mmdet/datasets/pipelines/transforms.py:2232]: 15 is a hard code here.

File: mmdet/datasets/pipelines/transforms.py [hunk 23]
```diff
@@ -575,20 +575,36 @@ class Pad:
...
+        pad2square (bool): Whether to pad the image into a square.
+           Currently only used for YOLOX. Default is False.
         pad_val (float, optional): Padding value, 0 by default.
     """
 
-    def __init__(self, size=None, size_divisor=None, pad_val=0):
+    def __init__(self,
+                 size=None,
+                 size_divisor=None,
+                 pad2square=False,
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:586]: pad_to_square

File: mmdet/datasets/pipelines/transforms.py [hunk 24]
```diff
@@ -575,20 +575,36 @@ class Pad:
     Args:
         size (tuple, optional): Fixed padding size.
         size_divisor (int, optional): The divisor of padded size.
+        pad2square (bool): Whether to pad the image into a square.
+           Currently only used for YOLOX. Default is False.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:579]: Default: False

File: mmdet/datasets/pipelines/transforms.py [hunk 25]
```diff
@@ -1908,6 +1926,458 @@ def __repr__(self):
...
+            center_ratio_range=(0.5, 1.5),
+            pad_value=114,
+    ):
+        assert isinstance(img_scale, tuple)
+        self.img_scale = img_scale
+        self.center_ratio_range = center_ratio_range
+        self.pad_value = pad_value
+
+    def __call__(self, results):
+        """Call function to mosaic image.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1982]: Call function to make a mosaic of image.

File: mmdet/datasets/pipelines/transforms.py [hunk 26]
```diff
@@ -1908,6 +1926,458 @@ def __repr__(self):
...
+            results (dict): Result dict.
+
+        Returns:
+            dict: Result dict with mosaic transformed.
+        """
+
+        results = self._mosaic_transform(results)
+        return results
+
+    def get_indexes(self, dataset):
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:1994]: Standard interface with "MultiImageMixDataset" for generating indexes

File: mmdet/datasets/pipelines/transforms.py [hunk 27]
```diff
@@ -1908,6 +1926,458 @@ def __repr__(self):
...
+        cropped sub-image.
+
+        Args:
+            loc (str): Index for the sub-image, loc in ('top_left',
+              'top_right', 'bottom_left', 'bottom_right').
+            center_position_xy (Sequence[float]): Mixing center for 4 images,
+                (x, y).
+            img_shape_wh (Sequence[int]): Width and height of sub-image
+
+        Returns:
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:2104]:
```
 Returns:
    tuple[tuple[float]]: Corresponding coordinate of pasting and croping
            
            - paste_coord (tuple): paste corner coordinate in mosaic image.
            - crop_coord (tuple): crop corner coordinate in mosaic image.
    ```

File: mmdet/datasets/dataset_wrappers.py [hunk 28]
```diff
@@ -280,3 +282,105 @@ def __getitem__(self, idx):
...
+    Suitable for training on multiple images mixed data augmentation like
+    mosaic and mixup. For the augmentation pipeline of mixed image data,
+    the `get_indexes` method needs to be provided to obtain the image
+    indexes, and you can set `skip_flags` to change the pipeline running
+    process. At the same time, we provide the `dynamic_scale` parameter
+    to dynamically change the output image size.
+
+    Args:
+        dataset (:obj:`CustomDataset`): The dataset to be mixed.
+        pipelines (Sequence[dict | callable]): Sequence of transform object or
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:300]: We use `pipline` in all dataset instead piplines

File: mmdet/datasets/dataset_wrappers.py [hunk 29]
```diff
@@ -280,3 +282,105 @@ def __getitem__(self, idx):
...
+        if skip_flags is not None:
+            assert len(skip_flags) == len(pipelines)
+            assert all(
+                [isinstance(skip_flag, bool) for skip_flag in skip_flags])
+        else:
+            skip_flags = [False] * len(pipelines)
+        self._skip_flags = skip_flags
+
+        self.pipelines = []
+        for pipeline in pipelines:
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:323]: Why not use `Compose`

File: mmdet/datasets/dataset_wrappers.py [hunk 30]
```diff
@@ -280,3 +282,105 @@ def __getitem__(self, idx):
...
+
+            if hasattr(pipeline, 'get_indexes'):
+                indexes = pipeline.get_indexes(self.dataset)
+                if not isinstance(indexes, collections.abc.Sequence):
+                    indexes = [indexes]
+                mix_results = [
+                    copy.deepcopy(self.dataset[index]) for index in indexes
+                ]
+                results['mix_results'] = mix_results
+                if self._dynamic_scale is not None:
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:359]: Please add comments for `img_scale` here

Comment (jshilong) [reply mmdet/datasets/dataset_wrappers.py:359]: The `img_scale` would be used in the :obj:`Resize` in the `self.pipeline`

File: mmdet/datasets/dataset_wrappers.py [hunk 31]
```diff
@@ -280,3 +282,105 @@ def __getitem__(self, idx):
...
+        if dynamic_scale is not None:
+            assert isinstance(dynamic_scale, tuple)
+        self._dynamic_scale = dynamic_scale
+
+    def __len__(self):
+        return self.num_samples
+
+    def __getitem__(self, idx):
+        results = self.dataset[idx]
+        for (pipeline, skip_flag) in zip(self.pipelines, self._skip_flags):
```
Comment (jshilong) [inline mmdet/datasets/dataset_wrappers.py:347]:
```
        for transform in self.pipeline:
            data = transform(data)
```

File: mmdet/datasets/pipelines/transforms.py [hunk 32]
```diff
@@ -1908,6 +1926,458 @@ def __repr__(self):
...
+            dict: Result dict with mixup transformed.
+        """
+
+        results = self._mixup_transform(results)
+        return results
+
+    def get_indexes(self, dataset):
+        """Call function to collect indexes.
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:2227]: Standard interface with :obj:`MultiImageMixDataset` for generating indexes

File: mmdet/datasets/builder.py [hunk 33]
```diff
@@ -66,6 +66,10 @@ def build_dataset(cfg, default_args=None):
     elif cfg['type'] == 'ClassBalancedDataset':
         dataset = ClassBalancedDataset(
             build_dataset(cfg['dataset'], default_args), cfg['oversample_thr'])
+    elif cfg['type'] == 'MultiImageMixDataset':
+        cfg['dataset'] = build_dataset(cfg['dataset'])
+        cfg.pop('type')
```
Comment (jshilong) [inline mmdet/datasets/builder.py:71]: Seem we can build the dataset in the `MultiImageMixDataset`

File: mmdet/datasets/pipelines/transforms.py [hunk 34]
```diff
@@ -1908,6 +1926,460 @@ def __repr__(self):
...
+                (results['gt_labels'], retrieve_gt_labels), axis=0)
+
+            results['img'] = mixup_img
+            results['gt_bboxes'] = mixup_gt_bboxes
+            results['gt_labels'] = mixup_gt_labels
+
+        return results
+
+    def _filter_box_candidates(self,
+                               box1,
```
Comment (jshilong) [inline mmdet/datasets/pipelines/transforms.py:2355]: - `box1` -> `bbox1`
- `wh_thr` -> `min_bbox_size`
- `ar_thr` -> `min_area_radio`
-  `aspect_ratio_thr` -> `max_aspect_ratio`

**Review by jshilong — APPROVED** [review]

**Comment by jshilong** [comment]
LGTM

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**